datacube/virtual/impl.py

```diff
@@ -0,0 +1,550 @@
+        grid_spec = index.products.get_by_name(self.product).grid_spec
+
+        return DatasetPile(datasets, grid_spec)
+
+    def group_datasets(self, datasets, **search_terms):
+        pile = datasets.pile
+        grid_spec = datasets.grid_spec
+
+        # select only those inside the ROI
+        # ROI could be smaller than the query for `find_datasets`
+        spatial_query = reject_keys(search_terms, self.NON_SPATIAL_KEYS)
+        polygon = query_geopolygon(**spatial_query)
+        selected = list(select_datasets_inside_polygon(pile, polygon))
+
+        # geobox
+        merged = merge_search_terms(select_keys(self.settings, self.NON_SPATIAL_KEYS),
+                                    select_keys(spatial_query, self.NON_SPATIAL_KEYS))
+        geobox_settings = select_keys(merged, self.GEOBOX_KEYS)
+
+        geobox = output_geobox(datasets=selected, grid_spec=grid_spec,
+                               **geobox_settings, **spatial_query)
+
+        # group by time
+        grouping_settings = select_keys(merged, self.GROUPING_KEYS)
+        grouped = Datacube.group_datasets(selected,
+                                          query_group_by(**grouping_settings))
+
+        def wrap(_, value):
+            return DatasetPile(value, grid_spec)
+
+        # information needed for Datacube.load_data
+        return GroupedDatasetPile(grouped, geobox).map(wrap)
+
+    def fetch_data(self, grouped, product_definitions, **load_settings):
+        assert isinstance(grouped, GroupedDatasetPile)
+
+        merged = merge_search_terms(select_keys(self.settings, self.LOAD_KEYS),
+                                    select_keys(load_settings, self.LOAD_KEYS))
+
+        # this method is basically `GridWorkflow.load`
+
+        # essentially what `datacube.api.core.set_resampling_method` does
+        if merged.get('resampling') is not None:
+            resampling = {'resampling_method': merged['resampling_method']}
+        else:
+            resampling = {}
+
+        # load_settings should not contain `measurements`
+        measurements = [Measurement(**measurement, **resampling)
+                        for _, measurement in self.output_measurements(product_definitions).items()]
+
+        def unwrap(_, value):
+            return value.pile
+
+        return Datacube.load_data(grouped.map(unwrap).pile,
+                                  grouped.geobox, measurements,
+                                  fuse_func=merged.get('fuse_func'),
+                                  dask_chunks=merged.get('dask_chunks'),
+                                  use_threads=merged.get('use_threads'))
+
+    def __repr__(self):
+        settings = ",".join("{}={}".format(key, value) for key, value in self.settings.items())
+        return "BasicProduct(product={}, {})".format(self.product, settings)
+
+
+class Transformation(ABC):
+    @abstractmethod
+    def measurements(self, input_measurements):
```

Comment:
Needs more docs

- Description of input output data structure
- Expected error handling behaviours for implementations (exception vs return None, that sort of stuff)